### ISSUE TITLE
Consider alternative default namespace approaches

### ISSUE THREAD

**Issue by timriley** [dry-rb/dry-system#23]
In implementing this feature and having to turn around the quick-fire fix in #22, I realised there might be some better ways to do this.

I was thinking that it might be better for the default namespace handling to be done on the _resolution_ side rather than the _registration_ side. dry-component could have its own resolver that is default namespace aware, so it would know to convert "foo" to "main.foo" if "main" was the default namespace and there was a matching registration.

For a convenience, this approach seems a lot cleaner and more efficient than getting involved in the actual file loading process. It also means all the container registrations remain "canonical" and fully matching their corresponding file paths, etc.

I was also thinking that it might be good if we can make it _explicit_ when a user intends to access a container registration as in the shorthand format (i.e. with the default namespace removed):

```ruby
# the leading `.` here implies the shorthand format
include Admin::Import[".users.repository"]
```

In this case, if `Admin::Import` is configured with "admin" as the default namespace, then it would know 100% that ".users.repository" should be converted to "admin.users.repository" and looked up accordingly.

What do you think? This feels like it would remove any possible ambiguity, which'd be a good thing.

**Comment by AMHOL** [comment]
I feel like the leading `.` should do the opposite and indicate that it should not be prefixed - same as the scope resolution operator (`::`).

**Comment by timriley** [comment]
@AMHOL interesting perspective! I had not thought of that.

I guess I see the leading dot acting kind of like how you chain methods across lines in Ruby:

```ruby
  foo(something)
    .bar
```

The ".bar" here, with it's leading dot, is _clearly_ "attached" to the item that preceded it. This is what I was going for with the shorthand container identifier syntax. WDYT?

**Comment by AMHOL** [comment]
I still think with the `.` being the namespace separator, it would be more intuitive to have it act as the scope-resolution operator. To imply that it is qualified with something would just confuse me personally.

**Comment by timriley** [comment]
I just realised I had `include Admin::Container[".users.repository"]` in my top description when I meant it to be `include Admin::Import[".users.repository"]`, i.e. the injector for a dry-component's container. I've corrected it now.

I have a feeling this correction wouldn't change your thoughts on this though, and that's fair enough, let's try and avoid anything that's potentially confusing.

So perhaps the best approach is to offer the same lookup as we offer now, with `"users.repository"` returning the object registered as `"admin.users.repository"` when `admin` is the component's default namespace.

This would mean that we still can try changing the namespace convenience lookups to be on the _resolution_ rather than the _registration_ side and we wouldn't have to change our public behaviour at all.
